### PR TITLE
[WIP]Add BinClash detecting build logger

### PR DESCRIPTION
**Pull request by ericstj**
This msbuild logger detects binclashes by recording all projects that
build and their TargetPath.  Once the build is done it logs all
clashes that occurred during the build.

This is a work in progress and I plan to refine the logging and
sequence of the error message.

/cc @weshaggard @mellinoe

### CODE REVIEW

**Comment by ericstj** [comment]
Here's sample output from the logger: https://gist.github.com/ericstj/e2be25ee073dfe7bd99d

It gets logged at the very end of the build, then I throw an exception to cause the build to fail.  It isn't great but its the only way I could find to have a logger fail the build.  I could write a file then have a target read the file and fail (emitting proper errors) if folks think that's better.

I could also emit the errors during the build at the point the project is building if folks like that better.

**Comment by mellinoe** [comment]
LGTM. We definitely need automatic validation for this, it is far too hard to get right by hand.

File: src/Microsoft.DotNet.Build.Tasks/BinClashLogger.cs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public class BinClashLogger : ILogger
+    {
+        internal class ProjectState
+        {
+            public ProjectState(ProjectStartedEventArgs args)
+            {
+                GlobalProperties = args.GlobalProperties;
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/BinClashLogger.cs]: Do you know if this is read-only at this point?

File: src/Microsoft.DotNet.Build.Tasks/BinClashLogger.cs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        }
+
+        /// <summary>
+        /// All projects evaluations that built, indexed by context ID.
+        /// </summary>
+        private Dictionary<int, ProjectState> _projectHistory = new Dictionary<int, ProjectState>();
+        private string _logFile = null;
+        private StreamWriter _fileWriter = null;
+
+        private bool _append = false;
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/BinClashLogger.cs]: Doesn't look like you are parsing _append.

File: src/Microsoft.DotNet.Build.Tasks/BinClashLogger.cs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            }
+
+            if (_fileWriter != null)
+            {
+                _fileWriter.Dispose();
+            }
+
+            if (_exceptionOnError && (clashes != 0))
+            {
+                throw new Exception($"{clashes} bin clashes were detected during the build.");
```
Comment (weshaggard) [inline src/Microsoft.DotNet.Build.Tasks/BinClashLogger.cs]: Is throwing an exception from the logger the right way to force a failure?

**Comment by weshaggard** [comment]
This is awesome! Thanks for digging into this and doing it, I should have dug into this approach earlier while writing my script.

File: src/Microsoft.DotNet.Build.Tasks/BinClashLogger.cs (hunk 3 above)
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/BinClashLogger.cs]: I spent quite a lot of time trying to find a way to gracefully error-out but I couldn't come up with anything.  I tried forwardinglogger/nodelogger which can let me generate new events but those events only flow to my central logger, not the default loggers.  I could have created some task that runs at the end of the build and communicate with it, but that felt less general purpose.

File: src/Microsoft.DotNet.Build.Tasks/BinClashLogger.cs (hunk 1 above)
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/BinClashLogger.cs]: It looks to me like its a detached copy of the properties, created with ToDictionary().  Also, this is a logger, it's purpose is to observe.  If it was possible to change the collection that'd be a bug in MSBuild and not something we should rely on.

File: src/Microsoft.DotNet.Build.Tasks/BinClashLogger.cs (hunk 2 above)
Comment (ericstj) [inline src/Microsoft.DotNet.Build.Tasks/BinClashLogger.cs]: Good catch.  I added it with the intent to make it configurable but forgot to parse the param string.

**Pull request merged**